### ISSUE TITLE
RetinaNet TPU Compatibility [looking for a long term fix instead of #1021]

### ISSUE THREAD

**Issue by LukeWood** [keras-team/keras-cv#1018]
We can enable this in `predict()`.

This might be a big performance delta, but is much lower priority than TPU compatibility.

**Comment by bhack** [comment]
Don't we want to use this ticket to investigate more https://github.com/keras-team/keras-cv/pull/1021#issuecomment-1320964584?

**Comment by LukeWood** [comment]
Sure, we can further investigate.

**Comment by bhack** [comment]
@ianstenbit if you can share your `gs://` pre-populated it could be useful as re-run the extraction is super-slow.

**Comment by ianstenbit** [comment]
@bhack sorry for the delay on this.
I've moved this data to `data_dir="gs://keras-cv/datasets"`, where you should have access. I've also updated my repro colab to reflect the change.

Also -- extraction on GCS hangs (see https://github.com/tensorflow/datasets/issues/4115), so make sure to pass `download=False` to your tfds.load call.

**Comment by bhack** [comment]
@ianstenbit Thanks

**Comment by bhack** [comment]
I don't think that the public url is `gs://keras-cv/datasets`. What we need to use?

**Comment by ianstenbit** [comment]
When I pull the public URL of e.g the dataset info, I am seeing this:

https://storage.googleapis.com/keras-cv/datasets/voc/2007/4.0.0/dataset_info.json

So I suppose you could try `https://storage.googleapis.com/keras-cv/datasets` -- but I would expect the gs version to work, so I'm not sure

**Comment by bhack** [comment]
What is the value of:
> Once public access has been granted, Copy URL appears for each object in the public access column. You can click this button to get the public URL for the object.

**Comment by ianstenbit** [comment]
Yes, I am able to reach it when I don't have an auth token (e.g. incognito mode)

Are you not able to reach those files in incognito mode?

**Comment by ianstenbit** [comment]
One other thing you can do is load+extract the VOC dataset locally and then upload the resulting tfds folder to a bucket that you have control over. (That's how I got these records in GCS in the first place, since extraction fails on TFDS.load for gcs)

**Comment by bhack** [comment]
> One other thing you can do is load+extract the VOC dataset locally and then upload the resulting tfds folder to a bucket that you have control over. (That's how I got these records in GCS in the first place, since extraction fails on TFDS.load for gcs)

Yes I know but it is faster if I can reuse your one.

https://storage.googleapis.com/keras-cv/datasets/voc/ in incognito mode:

```xml
<Error>
<Code>AccessDenied</Code>
<Message>Access denied.</Message>
<Details>
Anonymous caller does not have storage.objects.get access to the Google Cloud Storage object. Permission 'storage.objects.get' denied on resource (or it may not exist).
</Details>
</Error>
```

**Comment by ianstenbit** [comment]
Interesting. That works from my work device as well as my personal device.

I'm thinking that tfds.load probably requires object list permission in addition to access to the underlying files. I don't have the power to give that permission on any of these GCS buckets, so I'm not sure if there's anything else I can do.

FWIW you shouldn't need VOC specifically to repro the issue.
You could create a fake bbox dataset with [this method](https://github.com/keras-team/keras-cv/blob/master/keras_cv/models/object_detection/__test_utils__.py) and use that as train_ds and eval_ds.

If you'd like I can update the repro colab to use that instead of VOC

**Comment by bhack** [comment]
Thank you but I cannot find a free TPU anymore now..

**Comment by bhack** [comment]
Also on tf-nightly I still see:
```python
(0) INVALID_ARGUMENT: {{function_node __inference_train_function_133447}} TF to XLA legalization failed: <unknown>:0: error: loc(fused["SigmoidGrad:", "gradient_tape/FocalLoss/Sigmoid/SigmoidGrad"]): 'mhlo.constant' op result #0 must be statically shaped tensor of floating-point or pred (AKA boolean or 1-bit integer) or 8/16/32/64-bit signless integer or 8/16/32/64-bit unsigned integer or complex type with 32-bit float or 64-bit float elements values, but got 'tensor<?x?x20xf32>'
<unknown>:0: note: loc(fused["SigmoidGrad:", "gradient_tape/FocalLoss/Sigmoid/SigmoidGrad"]): see current operation: %5996 = "mhlo.constant"() {value = dense<1.000000e+00> : tensor<?x?x20xf32>} : () -> tensor<?x?x20xf32>
```



I cannot dump on the local filesystem or GCS:
```python
os.environ["GOOGLE_CLOUD_PROJECT"] = "tf-develop-350208"
os.environ["TF_DUMP_GRAPH_PREFIX"]= "gs://keras_cv/dataset/" # or /tmp/
os.environ["TF_XLA_FLAGS"]="--tf_xla_clustering_debug --tf_xla_auto_jit=2"
os.environ["XLA_FLAGS"]="--xla_dump_hlo_as_text --xla_dump_to=/tmp/generated"
```

@smit-hinsu How we could dump on Colab TPU (I still see https://github.com/tensorflow/tensorflow/issues/49702)?

**Comment by smit-hinsu** [comment]
Sorry, I don't have an idea on that.

**Comment by bhack** [comment]
@smit-hinsu Do you know someone that could help on this failed legalization?

**Comment by bhack** [comment]
We have a problem with:
https://github.com/tensorflow/tensorflow/commit/76079c00ac45b2bd48b3749a490e464149f4b244#diff-20844a653ebc1cb9b4b2adb26672259647f08dc1e78a06458d2e753a1449336bR2070

But in our case we have `tensor<?x?x20xf32>`  in the sigmoid grad where the constrain is `#0 must be statically shaped tensor `

https://github.com/keras-team/keras-cv/blob/d2d56d60650caa6d90fa74f3b91a111d2007d135/keras_cv/losses/focal.py#L81-L82

**Comment by bhack** [comment]
Tracking this upstream at https://github.com/tensorflow/tensorflow/issues/58645

**Comment by tanzhenyu** [comment]
Done through recent series of refactoring
